src/caffeine/caffeine.c

```diff
@@ -53,110 +54,98 @@ void c_sync_all()
+}
 
-      if (stat != NULL) *stat = 0;
+void c_co_min_double(void* c_loc_a, size_t Nelem, int* stat, int result_image)
+{
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_DBL, sizeof(double), GEX_OP_MIN, NULL, NULL);
 }
 
-void c_co_max_no_result_image_int32(void* c_loc_a, size_t Nelem, int* stat)
+void c_co_max_int32(void* c_loc_a, size_t Nelem, int* stat, int result_image)
 {
-      gex_Event_t ev
-        = gex_Coll_ReduceToAllNB(myteam, c_loc_a, c_loc_a, GEX_DT_I32, sizeof(int32_t), Nelem, GEX_OP_MAX, NULL, NULL, 0);
-      gex_Event_Wait(ev);  
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_I32, sizeof(int32_t), GEX_OP_MAX, NULL, NULL);
+}
 
-      if (stat != NULL) *stat = 0;
+void c_co_max_int64(void* c_loc_a, size_t Nelem, int* stat, int result_image)
+{
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_I64, sizeof(int64_t), GEX_OP_MAX, NULL, NULL);
 }
 
-void c_co_max_no_result_image_int64(void* c_loc_a, size_t Nelem, int* stat)
+void c_co_max_float(void* c_loc_a, size_t Nelem, int* stat, int result_image)
 {
-      gex_Event_t ev
-        = gex_Coll_ReduceToAllNB(myteam, c_loc_a, c_loc_a, GEX_DT_I64, sizeof(int64_t), Nelem, GEX_OP_MAX, NULL, NULL, 0);
-      gex_Event_Wait(ev);  
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_FLT, sizeof(float), GEX_OP_MAX, NULL, NULL);
+}
 
-      if (stat != NULL) *stat = 0;
+void c_co_max_double(void* c_loc_a, size_t Nelem, int* stat, int result_image)
+{
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_DBL, sizeof(double), GEX_OP_MAX, NULL, NULL);
 }
 
-void c_co_max_no_result_image_float(void* c_loc_a, size_t Nelem, int* stat)
+void c_co_reduce_int32(void* c_loc_a, size_t Nelem, int* stat, int result_image, gex_Coll_ReduceFn_t* operation)
 {
-      gex_Event_t ev
-        = gex_Coll_ReduceToAllNB(myteam, c_loc_a, c_loc_a, GEX_DT_FLT, sizeof(float), Nelem, GEX_OP_MAX, NULL, NULL, 0);
-      gex_Event_Wait(ev);  
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_I32, sizeof(int32_t), GEX_OP_USER, operation, NULL);
+}
 
-      if (stat != NULL) *stat = 0;
+void c_co_reduce_float(void* c_loc_a, size_t Nelem, int* stat, int result_image, gex_Coll_ReduceFn_t* operation)
+{
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_FLT, sizeof(float), GEX_OP_USER, operation, NULL);
 }
 
-void c_co_max_no_result_image_double(void* c_loc_a, size_t Nelem, int* stat)
+void c_co_reduce_char(void* c_loc_a, size_t Nelem, int* stat, int result_image, gex_Coll_ReduceFn_t* operation, void* client_data)
 {
-      gex_Event_t ev
-        = gex_Coll_ReduceToAllNB(myteam, c_loc_a, c_loc_a, GEX_DT_DBL, sizeof(double), Nelem, GEX_OP_MAX, NULL, NULL, 0);
-      gex_Event_Wait(ev);  
+     c_co_reduce_universal(c_loc_a, Nelem, stat, result_image, GEX_DT_USER, sizeof(char), GEX_OP_USER, operation, client_data);
```

Comment:
I think this is a bug. 

IIUC this is a user-defined operator reduction where each rank passes a character string with fixed length L, and as written ends up calling `gex_Coll_ReduceTo{One,All}NB(GEX_DT_USER,GEX_OP_USER)` with dt_sz==sizeof(char) and dt_cnt=L, but this is **probably wrong for character strings**. Specifically, we are not reducing an array of L independent 1-byte elements that GASNet is free to fragment and invoke the callback on smaller pieces (which I believe it WILL do, for sufficient large dt_cnt). Rather, this is a reduction over an L-byte datatype where each string is an indivisable datum as far as callbacks are concerned. So dt_sz  should be L, and dt_cnt should be 1 (unless/until you add support for an array of strings).

CC: @phhargrove

Reply:
This comment is now linked in issue #18

Reply:
@bonachea I have verified that this issue has been resolved in PR #21.  As of the current head of the corresponding branch, the `co_min`, `co_max`, and `co_reduce` tests employ a range of string lengths and string array shapes.  In each case, the call to `gex_Coll_ReduceTo{One,All}` results in `dt_sz` representing the the number of bytes in a  single string and `dt_cnt` being the number of strings that each rank holds in its string array (with `dt_cnt` being 1 when each rank just holds scalar, i.e., one `dt_sz`-byte string).  All tests pass, including tests with scalars, 1D, 2D, and 3D arrays, including one test in which the array size and element length are both set at runtime.  I'll close issue #18.